[요구 기능 추가] 캐싱, 페이징 추가, 구글 체크 스타일 적용 + 최신순, 좋아요순 조회 기능 추가

thinker-94 opened the pull request.

thinker-94 committed:
[기능 추가] 사용자가 책 카테고리를 클릭하면 나오는 데이터 캐싱

thinker-94 committed:
[유지보수] 책 조회시 NULL값 나오는 문제 해결

thinker-94 committed:
[기능 추가, 코드 리팩토링] 페이징 기능 추가 / json 직렬화 코드 추가 / response header 수정

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/JsonUtils.java:
```diff
@@ -0,0 +1,33 @@
+    private JsonUtils() {}
+
+    public static String makeBookListToJson(List<Book> bookList) throws JsonProcessingException {
+        ObjectMapper objectMapper = new ObjectMapper();
```
`ObjectMapper` 클래스는 `Thread safe`하기 때문에 한번만 생성해놓고 여러 스레드에서 돌려써도 됩니다.
현재는 메소드를 호출할때마다 이 객체를 새로 생성하고 있는데요, 이 클래스에 대해 알아보시고 한번만 생성하고 돌려쓸 수 있도록 해주시면 좋겠습니다~

youngsuk-dev replied:
- [ ]

youngsuk-dev replied:
[문제 해결 방법]
check style을 코드에 지속적으로 적용하기 위해 ide에 플러그인을 설치 했습니다.

[참고 자료]
https://medium.com/@somindagamage/how-to-configure-checkstyle-and-findbugs-plugins-to-intellij-idea-38148aad2387

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/JsonUtils.java:
```diff
@@ -0,0 +1,33 @@
+    // 해당 메소드가 없게 되면 리스트 형태의 데이터들이 어떤 값인지 직관적으로 알기 어려운 문제가 생기게 됩니다.
+    private static Map<String, List<Book>> makeBookListToHashMap(List<Book> bookList) {
+        Map<String, List<Book>> bookListHashMap = new HashMap<>();
+        bookListHashMap.put("bookList",bookList);
```
`,` 다음에 공백을 넣어주면 좋겠습니다. 기술적인 것도 중요하지만 이런 코딩 컨벤션을 꼼꼼하게 지켜주시는게 좋습니다~
checkstyle을 적용하면 더 좋을 것 같고요

thinker-94 replied:
[문제 해결 방법]
google check style을 코드에 지속적으로 적용하기 위해 ide에 플러그인을 설치 했습니다. check style은 다른 기술들과 달리 경험하는 것 만으로도 의미가 있다고 생각 했기 때문에 google check style을 적용했습니다.

앞으로 프로젝트를 진행할때 google check style을 지속적으로 준수 하겠습니다.

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/JsonUtils.java:
```diff
@@ -0,0 +1,33 @@
+    //책들을 json 형태로 클라이언트에게 전달하는 api를 만들때 가독성을 높이기 위해 추가한 메소드 입니다.
+    //이 메소드를 통해서 bookList라는 스트링값 안에 리스트 형태의 책들이 json에 들어가게 됩니다.
+    // 해당 메소드가 없게 되면 리스트 형태의 데이터들이 어떤 값인지 직관적으로 알기 어려운 문제가 생기게 됩니다.
+    private static Map<String, List<Book>> makeBookListToHashMap(List<Book> bookList) {
```
클래스 이름은 `JsonUtils`인데 `Book` 도메인과 관련된 로직이 이 클래스에 들어있네요.
클래스의 책임에 벗어나는 행위가 아닐까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/PagingUtils.java:
```diff
@@ -0,0 +1,14 @@
+
+
+    public static int makeOffsetForPaging(int currentPageNumber) {
+        int onePageContentAmount = 10;
```
이 값은 상수로 사용하면 좋을 것 같습니다~ 혹은 외부에서 설정을 주입할 수 있으면 좋지 않을까요? 페이징 관련 값들은 바뀔 가능성이 있으니까요

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/etc.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class etc {
```
- 클래스의 작명규칙에 맞지 않습니다~ 카멜케이스를 지켜주세요
- etc라는 이름은 역할이 애매한것들을 여기다 모아두겠다는 의도로 보입니다. 클래스의 역할은 명확해야합니다~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/etc.java:
```diff
@@ -0,0 +1,19 @@
+    //2개의 파라미터를 넘기고 싶었지만 방법이 Map을 사용해서 하는 방법과, POJO를 사용해서 넘기는 방법이 있었다.
+    //POJO를 사용하면 하나의 메소드에 있는 파라미터 뿐만 아니라 또 다른 메소드의 파라미터까지 바꿔야 하는 문제가 있어서 Map을 사용했다.
+    public static Map makeParamToHashmap(String categoryName, int offsetForPaging) {
+        Map map = new HashMap();
```
타입이 명확해지므로 제네릭은 명시해주는게 좋습니다~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/etc.java:
```diff
@@ -0,0 +1,19 @@
+    private etc() {}
+
+    //2개의 파라미터를 넘기고 싶었지만 방법이 Map을 사용해서 하는 방법과, POJO를 사용해서 넘기는 방법이 있었다.
+    //POJO를 사용하면 하나의 메소드에 있는 파라미터 뿐만 아니라 또 다른 메소드의 파라미터까지 바꿔야 하는 문제가 있어서 Map을 사용했다.
```
- 주석의 `//` 이후에 공백을 하나 넣어주는게 가독성이 좋습니다
- 주석에서 배경과 이유가 명확하게 이해되지 않네요~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/BookController.java:
```diff
@@ -1,27 +1,38 @@
-
-@RestController
+@Controller
+@RequestMapping(value = "/books/category/", produces = "application/json; charset=UTF-8" )
```
`category`도 복수형으로 넣어주시는게 좋을 것 같네요

thinker-94 replied:
[리뷰에 대한 생각]
category 보다, categorys를 사용하는것이 좋겠네요
제 API를 사용하는 사람의 입장으로 생각했을때, books/category/ 의 의미는 카테고리 하나만 가져와줘 라는 말처럼 들리니까요

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/BookController.java:
```diff
@@ -1,27 +1,38 @@
+
+
+
+
```
굳이 필요없는 공백으로 보입니다~

thinker-94 replied:
[문제 해결 방법]
공백 제거

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/BookController.java:
```diff
@@ -1,27 +1,38 @@
-
+    @GetMapping(value = "{categoryName}")
+    public ResponseEntity<String> getBookByCategory(@PathVariable String categoryName,
+                                                    //currentPageNumber값에 아무것도 들어있지 않으면 첫번째 페이지를 리턴해주도록 defaltValue를 1로 설정 하였습니다
```
👍

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -1,21 +1,46 @@
+    /***
+     *
+     * [@Cacheable 동작 방식]
+     * 1. 어노테이션의 key값을 캐시 저장소에서 확인한다.
```
`캐시 저장소`란 어떤 것을 의미할까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -1,21 +1,46 @@
+        //그런데 직렬화의 반대의 과정인 역직렬화에는 보안적인 문제가 존재한다.
+        //이유는 역직렬화 과정에 해커가 악의적인 데이터를 집어넣을 수 있기 떄문이다.
+        //이 문제는 자바 직렬화, jackson 등의 여러 직렬화 기술들에서 발견되고 있다.
+        //나는 현재 역직렬화 과정은 거치지 않으므로 대안에 대해서는 아직 깊이 검토하지는 않았다.
```
레디스를 캐시저장소로 사용한다면 캐시에서 가져올때 역직렬화를 거치지 않을까요?

f-lab-dev commented on src/main/resources/query/user/book/bookQuery.xml:
```diff
@@ -10,17 +10,19 @@
         from book left join author on book.id = author.id
         where book_category = (#{bookCategoryName})
-        order by id DESC LIMIT 10;
+        order by id DESC LIMIT 10 OFFSET #{offset}
```
offset 기반 페이징은 성능적인 문제가 있을 수 있습니다. 이 페이징 방식을 선택하신 이유가 있으신가요?

f-lab-dev commented on src/main/resources/query/user/book/bookQuery.xml:
```diff
@@ -10,17 +10,19 @@
-    <select id="selectBookByCategory" resultMap="DBNameMappingBookDTO">
-        select book.id, book.book_name, author.author_name, book.book_publish_date
+    <select id="selectBookByCategory" parameterType="map" resultMap="DBNameMappingBookDTO">
+        select book.id, book.book_name, book.book_category, author.author_name, book.book_publish_date,
```
예약어는 대문자로 써준다거나, 컬럼마다 개행을 넣어준다거나 등 쿼리 가독성도 신경써주시면 좋겠습니다~

thinker-94 committed:
[유지보수] 메소드 가독성을 위해서 제네릭 타입 추가

thinker-94 committed:
구글 체크 스타일 기존 코드 전체에 적용

thinker-94 committed:
코드 리뷰 반영과 최신순 좋아요 순 정렬 & 유저가 페이지 수를 설정할 수 있도록 기능 추가 (페이지 수 설정 제한은 추가할 예정)

f-lab-dev commented:
일단 한 브랜치에 작업 내용이 너무 많습니다. 한 브랜치는 한 이슈 개발만을 해야하며 이슈는 최대한 잘게 분리해야합니다~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/JsonUtils.java:
```diff
@@ -11,23 +11,23 @@
-        ObjectMapper objectMapper = new ObjectMapper();
-        Map<String, List<Book>> bookListHashMap = makeBookListToHashMap(bookList);
+  public static String makeBookListToJson(List<Book> bookList) throws JsonProcessingException {
+    ObjectMapper objectMapper = new ObjectMapper();
```
굳이 매번 `ObjectMapper` 클래스의 객체를 생성해주어야할까요? `ObjectMapper`는 `thread safe`하기 때문에 하나만 생성하고 돌려서 사용해도 됩니다.

관련 내용은 찾아보시면 좋겠네요~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/JsonUtils.java:
```diff
@@ -11,23 +11,23 @@
+  //책들을 json 형태로 클라이언트에게 전달하는 api를 만들때 가독성을 높이기 위해 추가한 메소드 입니다.
+  //이 메소드를 통해서 bookList라는 스트링값 안에 리스트 형태의 책들이 json에 들어가게 됩니다.
+  // 해당 메소드가 없게 되면 리스트 형태의 데이터들이 어떤 값인지 직관적으로 알기 어려운 문제가 생기게 됩니다.
+  private static Map<String, List<Book>> makeBookListToHashMap(List<Book> bookList) {
```
1. 이 클래스의 이름은 `JsonUtils`입니다. 그러면 json과 관련된 로직만 들어가야하는데 책과 관련된 로직이 들어있습니다.
2. 결국 이 메소드의 역할은 리스트에 이름을 달아주기 위해서 존재하는 것으로 보입니다. 보통 책의 리스트를 조회할때는 `/books` 같이 URL만으로 **책의 리스트를 조회한다** 라는 기능이라는 것을 유추가 가능한데 굳이 이렇게 이름을 달아줘야할까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/LoginResponseUtils.java:
```diff
@@ -1,40 +1,41 @@
-    }
+  //사용자들이 로그인을 할때 객체 생성 하는 횟수가 많아지게 하는것보다
+  //static을 통해서 객체 생성 횟수를 줄이는 방법이 효율적이라 생각해서 static으로 만듬
+  public static User makeUserAddResponseInformation(User user) {
```
이 메소드의 역할은 유저의 패스워드를 마스킹해주는 역할입니다. 이름을 적절하게 다시 지어주면 좋겠네요

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/LoginResponseUtils.java:
```diff
@@ -1,40 +1,41 @@
 
-        return loginMessage;
+    if (isLoginSuccess) {
+      loginMessage = loginSuccessMessage;
```
그냥 바로 리턴해주는게 더 간결하지 않을까요? 바로 리턴해주면 바로 아래의 else도 필요없어보입니다~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/PagingUtils.java:
```diff
@@ -2,13 +2,12 @@
-        //offset을 만드는 알고리즘 입니다.
-        //현재 페이지에서 1을 빼고 하나의 페이지에 들어갈 아이템의 갯수를 곱해서 다음 페이지의 시작점을 알 수 있습니다.
-        return (currentPageNumber -1) * onePageContentAmount;
+  public static int makeOffsetForPaging(int currentPageNumber) {
```
offset 방식 페이징과 cursor 방식 페이징은 어떤 차이가 있을까요?

thinker-94 replied:
만약 총 1000페이지 분량의 데이터가 있을때 마지막 페이지를 사용자에게 주기 위해 db에 쿼리를 날린다면, offset 방식 페이징은 모든 row를 확인(row look up)을 합니다. 반면 cursor는 조건을 걸기 때문에 row look up을 offset에 비해 최소화 하는 차이가 있습니다. 

데이터베이스에 데이터가 굉장히 많은 상황이라면 offset 보다는 cursor 기반 페이징을 사용하는게 유리할 수 있습니다.

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/config/RedisConfiguration.java:
```diff
@@ -10,28 +10,27 @@
- */
+
+
+//[세션 공부]
```
여기는 다중행 주석에서 단일행 주석으로 바꾸셨네요. 다중행 주석을 사용하는게 더 맞지 않을까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/BookController.java:
```diff
@@ -13,26 +13,17 @@
 
 @Controller
-@RequestMapping(value = "/books/category/", produces = "application/json; charset=UTF-8" )
+@RequestMapping(value = "/books/categorys/", produces = "application/json; charset=UTF-8" )
```
categories 오타같네요~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/UserController.java:
```diff
@@ -12,54 +13,51 @@
+  }
+
+  @PostMapping(path = "/users/login")
+  public ResponseEntity<User> userCheckPasswordGiveSession(User user, HttpServletRequest request) {
```
`login`이라는 이름으로 충분해보이는데 이름이 너무 복잡하네요~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/UserController.java:
```diff
@@ -12,54 +13,51 @@
+
+  private void setUserSession(User user, HttpServletRequest request) {
+    HttpSession session = request.getSession();
+    session.setAttribute("userId",user.getUserId());
```
세션 키는 이렇게 하드코딩 되어있으면 관리하기 어렵습니다~ 따로 세션키를 관리하는 클래스를 두면 어떨까요?

thinker-94 replied:
사용자가 많아지면 사용자의 권한별로 세션키값을 다르게 할 수 있기 때문 일까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -15,32 +15,29 @@
+    //이 문제는 자바 직렬화, jackson 등의 여러 직렬화 기술들에서 발견되고 있다.
+    //나는 현재 역직렬화 과정은 거치지 않으므로 대안에 대해서는 아직 깊이 검토하지는 않았다.
+    //지금 까지 찾아본 대안으로는 커스텀한 직렬화 방법을 만드는 것이다.
+    //이 방법을 사용하면 해커가 데이터의 구조를 파악하기 어렵기 때문에 역직렬화 과정에서 악의적인 데이터를 주입하기가 어려워진다.
```
직렬화 주석은 왜 여기 달려있는걸까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/UserInformationService.java:
```diff
@@ -9,24 +9,26 @@
+  @Autowired
+  private UserRepository userRepository;
+
+  public User makeUserPasswordEncrypt(User user) {
```
**비밀번호를 암호화한다**라는 로직은 **유저 추가**라는 기능의 일부입니다. 메소드명에 굳이 드러낼 필요는 없어보이네요~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/UserInformationService.java:
```diff
@@ -9,24 +9,26 @@
+    return user;
+  }
+
+  public void insertUserData(User user) {
```
암호화해서 insert한다, 그냥 insert한다 굳이 이렇게 두개를 따로 제공할 필요가 있을까요?
각 메소드는 어떨 때 따로 쓰이길래 둘 다 개발하신걸까요?

thinker-94 replied:
비밀번호 암호화와 유저 회원가입 정보 저장은 같이 동작하기 때문에 메소드를 분리할 필요가 없겠네요

f-lab-dev commented on pom.xml:
```diff
@@ -66,6 +66,14 @@
+        <!-- SQL문 로그를 확인하기 위해 넣은 라이브러리 -->
+        <dependency>
+            <groupId>org.bgee.log4jdbc-log4j2</groupId>
+            <artifactId>log4jdbc-log4j2-jdbc4.1</artifactId>
```
mybatis는 SQL로그를 지원하지 않나요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/BookListToJsonUtils.java:
```diff
@@ -0,0 +1,26 @@
+
+  private static final Map<String, List<Book>> useOfPutListInJsonKeyStringHashMap = new HashMap<>();
+
+  public static String putBookListInJsonKeyString(List<Book> bookList, ObjectMapper objectMapper)
```
꼭 그때그때 objectMapper를 주입해줘야할까요? 이 메소드를 사용하는 사용자가 불편할 것 같습니다

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -1,43 +1,61 @@
+    book.setOnePageLimit(onePageLimit);
+
+    List<Book> bookList;
+    String bookListKeyName = "bookList";
```
이 값은 고정되어있는데 전역 상수로 사용하는게 낫지 않을까요?

f-lab-dev commented on src/main/resources/query/user/book/bookQuery.xml:
```diff
@@ -16,13 +16,41 @@
+        author.author_name, book.book_publish_date,
+        book.book_introduce, author.author_introduce,
+        book.book_index_text , book.book_cost, book.book_like_count, book.book_buy_count
+
```
이 쿼리 전반적으로 개행이 과도해보입니다. 들여쓰기를 활용하는게 낫지 않을까요?

f-lab-dev commented on src/main/resources/query/user/book/bookQuery.xml:
```diff
@@ -16,13 +16,41 @@
+        FROM
+        book
+
+        LEFT JOIN
```
`LEFT JOIN`과 `INNER JOIN`의 차이는 무엇일까요?

f-lab-dev commented on src/main/resources/query/user/book/bookQuery.xml:
```diff
@@ -16,13 +16,41 @@
+            book.book_publish_date
+        </if>
+
+        <if test='sortType.equals(Likes)'>
```
`sortType`이 위의 두 값에 해당되지 않으면 잘못된 SQL이 생성되어 에러가 발생될 것 같습니다. 문자열로 만들고 상수화시키는것보단 enum을 활용해보는건 어떨까요?

thinker-94 committed:
코드 리뷰 반영 완료

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/LoginResponseUtils.java:
```diff
@@ -9,31 +9,20 @@
-  public static User makeUserAddResponseInformation(User user) {
-    user.setUserPassword("*********");
-    user.setUserMessage(userAddSuccessMessage);
+  /* 사용자들이 로그인을 할때 객체 생성 하는 횟수가 많아지게 하는것보다
```
별로 주석이 예쁘지 않은데 아래처럼 해보면 어떨까요?

```java
/*
    내용
*/
```

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/LoginResponseUtils.java:
```diff
@@ -9,31 +9,20 @@
-    } else {
-      loginMessage = loginFailureMessage;
-    }
+  public static UserDTO maskingUserPassword(UserDTO userDTO) {
```
UserDTO 타입을 리턴한다면 메소드를 사용하는 입장에서는 인자로 받는 `userDTO`가 바뀌지 않을 것이라고 생각할 수도 있습니다.
`immutable`을 지켜주기 위해 setter를 없애고 패스워드와 메세지만 바뀐 새 객체를 만들어서 리턴해주는건 어떨까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/constants/SessionKeyConstants.java:
```diff
@@ -0,0 +1,8 @@
+public class SessionKeyConstants {
+  private SessionKeyConstants() {}
+
+  public static final String levelOne = "levelOne";
```
`levelOne`이라는 변수명으로는 어떤 역할인지 알 수가 없네요. 이름을 더 신경써서 지어주시고 역할에 대한 적절한 주석도 넣어주세요~
그리고 상수명은  `LEVEL_ONE` 처럼 지어야합니다

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/config/RedisConfiguration.java:
```diff
@@ -12,43 +12,53 @@
+생성된 세션은 웹브라우저의 쿠키에 저장된다.
+웹서버는 웹브라우저의 쿠키를 통해서 사용자의 로그인을 유지 할 수 있다.
+세션에는 유효 기간 설정을 해줘야 한다. 왜냐면 사용자가 많이 들어와서
+세션 객체가 계속 생성된다면 메모리가 부족해질 수 있기 때문이다
```
주석을 공부 노트로 사용하는거는 너무 과도한 것 같고, 자신이 사용하는 어노테이션이나 설정값에 대한 내용만 적어주세요~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/config/RedisConfiguration.java:
```diff
@@ -12,43 +12,53 @@
+/*
+로그인한 사용자가 결제 페이지에 머무르는 시간이 가장 많을것이라고 생각했다.
+사용자가 이사를해서 주소를 새로 입력하거나 핸드폰을 바꿔서 번호를 바꿀 수도 있기 때문이다.
+결제를 할때도 시간이 걸릴것이라고 생각했다. 이런 시간들을 고려했을때 10분이면 적절할것이라고 생각했다.
```
결제를 하면서 뭔가 정보를 잊어버린다거나 혹은 기타 결제방식에 대한 문제가 발생했을 때를 가정하면 10분은 짧아보입니다~ 특별한 사유가 없다면 기본값을 사용해도 됩니다

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/UserController.java:
```diff
@@ -1,63 +1,64 @@
 
-      return ResponseEntity.status(HttpStatus.OK).body(user);
+    if (loginSuccess) {
+      HttpSession session = request.getSession();
```
메소드에서 `HttpServletRequest` 대신 `HttpSession`을 주입받을 수도 있습니다

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/dto/BookDTO.java:
```diff
@@ -0,0 +1,29 @@
+
+@Getter
+@Setter
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
이 어노테이션을 달아준 이유가 무엇인가요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/dto/BookDTO.java:
```diff
@@ -0,0 +1,29 @@
+import lombok.Setter;
+
+@Getter
+@Setter
```
꼭 setter가 있어야할까요? 거의 모든 DTO에 다 들어가고있던데 `immutable`에 대해 더 알아보시면 좋을 것 같아요~ 왜 그래야하는지도 알아보시면 좋을 것 같고요

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/repository/UserRepository.java:
```diff
@@ -1,51 +1,47 @@
+  - 비즈니스 로직을 처리하는 역할을 한다. 예를들어 시청 인원수를 올린다던가, 결제를 처리한다던가 하는 역할을 한다.
+  - Service layer가 필요한 이유는 관심사를 분리해서 유지보수를 편하게 할 수 있기 위해서다.
+3. persistance layer (@Repository 사용)
+  - 데이터를 저장하고 삭제하고 수정하는등의 역할을 한다.
```
이 내용은 좋습니다~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -20,42 +21,23 @@
-    return putBookListInJsonKeyString(bookList, objectMapper);
+   /*
+   [@Cacheable 동작 방식]
+   1. key값을 Redis에서 확인한다. (redis-cli monitor를 통해서 확인)
```
`@Cacheable`은 무조건 redis에 캐싱하나요? 어떤 조건일 때 레디스에 캐싱하는지 더 깊게 알아보시면 좋겠습니다

thinker-94 replied:
넵 깊게 알아보겠습니다

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -20,42 +21,23 @@
+    bookDTO.setSortTypeEnum(sortTypeEnum);
+    List<BookDTO> bookList = bookRepository.selectBookByCategory(bookDTO);
+
+    return objectMapper.writeValueAsString(bookList);
```
직렬화를 직접 해주는 이유가 있나요? 어떤 설정을 하면 이렇게 직접 직렬화해주지 않아도 자동으로 직렬화를 해줄 수 있습니다.

thinker-94 committed:
코드리뷰 반영 및 에러 해결

thinker-94 committed:
코드 리뷰 반영 완료

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/common/utils/constants/SessionKeyConstants.java:
```diff
@@ -3,6 +3,6 @@
   private SessionKeyConstants() {}
 
-  public static final String levelOne = "levelOne";
+  public static final String LEVEL_ONE_SESSION_KEY = "levelOne";
```
level one 이라고 하면 어떤 권한을 갖고있는 사람인지 명확하지 않습니다. user, admin 등의 이름을 써주는게 명확할 것 같네요~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/config/RedisConfiguration.java:
```diff
@@ -73,8 +49,7 @@
   public RedisConnectionFactory lettuceConnectionFactory() {
     RedisStandaloneConfiguration redisStandaloneConfiguration = new RedisStandaloneConfiguration(redisHostName, redisPort);
```
여기엔 `Standalone` config를 사용했습니다. 다른 config도 있을까요?

thinker-94 replied:
1. RedisClusterConfiguration
2. RedisSentinelConfiguration
3. RedisSocketConfiguration
4. RedisStaticMasterReplicaConfiguration

어떤 환경으로 redis에 연결하는지에 따라 다른 config들이 있습니다. 저는 하나의 node의 redis만 사용하기 때문에 standalone을 사용하고 있습니다

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/UserController.java:
```diff
@@ -36,25 +34,24 @@
 
-  @PostMapping(path = "/users/login")
-  public ResponseEntity<UserDTO> login(UserDTO userDTO, HttpServletRequest request) {
+  @PostMapping(path = "/login")
```
`path = ` 이거는 생략해줄 수 있지 않을까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/dto/BookDTO.java:
```diff
@@ -20,6 +18,18 @@
   private String bookIndexDescription;
   private String bookLikeCount;
 
+  public void setBookCategory(String bookCategory) {
```
왜 setter 어노테이션을 없애고 직접 만드셨나요? 기존과 동일해보이네요

thinker-94 replied:
멤버변수에는 @setter를 사용할 수 있는지 몰라서 그렇습니다. 코드 수정 하였습니다

thinker-94 replied:
> 왜 setter 어노테이션을 없애고 직접 만드셨나요? 기존과 동일해보이네요

멤버변수에는 @Setter를 사용할 수 있는지 몰라서 그렇습니다. 코드 수정 하였습니다

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -18,26 +16,22 @@
+  */
+
+  //TODO. key 추가로 만들것
+  @Cacheable(cacheNames = "BookByCategoryCache", key = "{#categoryName}")
```
- `Cacheable`을 사용하는데 어떤 과정을 거쳤길래 redis가 캐시 저장소로 사용될까요? 기본 값은 어떤 것일까요?
- 그리고 다시 강조드리지만 잘게 나눈 1개의 이슈 = 1개의 브랜치 = 1개의 PR 입니다~ 다음에는 이슈를 잘게 나눠서 모두 다 다른 브랜치로 작업해주세요~

thinker-94 replied:
> * `Cacheable`을 사용하는데 어떤 과정을 거쳤길래 redis가 캐시 저장소로 사용될까요? 기본 값은 어떤 것일까요?
> * 그리고 다시 강조드리지만 잘게 나눈 1개의 이슈 = 1개의 브랜치 = 1개의 PR 입니다~ 다음에는 이슈를 잘게 나눠서 모두 다 다른 브랜치로 작업해주세요~

RedisConnectionFactory, StringRedisTemplate, RedisTemplate을 빈으로 등록하면 Spring boot 의 @EnableAutoConfiguration 어노테이션을 통해서 Redis를 찾습니다. ConcurrentHashMap이 기본 캐시 저장소입니다.

spring-boot-autoconfigure jar 파일에 있는 외부설정 파일에 스프링 부트에 있는 모든 설정들이 특정 조건에 의해 정해집니다. 캐싱도 마찬가지입니다.

추가로 @EnableCaching은  스프링 빈을 모두 검사해서 @Cacheable과 같은 캐시 관련된 어노테이션이 달려있는 빈의 메소드를 찾습니다.

f-lab-dev reviewed (commented):
커밋 메세지를 보면 `코드리뷰 반영 완료` 처럼 되어있는데 이 메세지만 봐서는 어떤 작업을 했는지 명확하지 않습니다.
커밋 메세지의 제목만 사용하지 마시고 내용도 달아주세요~

커밋메세지를 어떻게 작성해야하는지는 검색해보시면 나올겁니다

thinker-94 committed:
refactor: 모든 DTO 인스턴스의 상태가 최대한 변하지 않도록 수정

thinker-94 committed:
style: 컨벤션에 맞게 코드 수정, 주로 클래스명에 DTO 들어간 것을 Dto로 변경

thinker-94 commented:
> 커밋 메세지를 보면 `코드리뷰 반영 완료` 처럼 되어있는데 이 메세지만 봐서는 어떤 작업을 했는지 명확하지 않습니다.
> 커밋 메세지의 제목만 사용하지 마시고 내용도 달아주세요~
> 
> 커밋메세지를 어떻게 작성해야하는지는 검색해보시면 나올겁니다

넵

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/dto/UserDto.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.Setter;
+
+@Getter
+@Setter
```
이 클래스는 immutable을 깨는 이유가 무엇일까요? 적절한 이유를 주석으로 달아주세요~

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/dto/BookDto.java:
```diff
@@ -3,35 +3,29 @@
 
 @Getter
-public class BookDTO {
+public final class BookDto {
```
어떤 것은 final을 달고 어떤 것은 안달고있습니다. 목적에 맞춰 동일한 목적의 클래스는 똑같이 맞춰주세요

thinker-94 replied:
스프링에서 호출되고 있는 클래스들을 제외하고 모든 클래스들을 final을 달아주겠습니다. 객체의 데이터가 최대한 바뀌지 않게 하려면 클래스가 상속되지 않도록 해야되기 때문입니다. 그리고 spring은 cglib를 통해서 클래스를 상속시킨뒤 프록시를 생성하기 때문에 스프링에 의존하는 클래스들은 final을 붙일 수 없습니다. 만약 클래스에 상속이 필요하다면 그때 final을 제거하겠습니다. 클래스를 상속으로는 일반적으로 만들지 않다고 생각하기 때문입니다.

thinker-94 replied:
스프링에서 호출되고 있는 클래스들을 제외하고 모든 클래스들을 final을 달아주겠습니다. 객체의 데이터가 최대한 바뀌지 않게 하려면 클래스가 상속되지 않도록 해야되기 때문입니다. 그리고 spring은 cglib를 통해서 클래스를 상속시킨뒤 프록시를 생성하기 때문에 스프링에 의존하는 클래스들은 final을 붙일 수 없습니다. 만약 클래스에 상속이 필요하다면 그때 final을 제거하겠습니다. 클래스를 상속으로는 일반적으로 만들지 않다고 생각하기 때문입니다.

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -25,13 +26,16 @@
+    bookDto.setPagingDto(pagingDto);
+    bookDto.setSortTypeEnum(sortTypeEnum);
+
+
```
여기는 필요없는 개행이 꽤 많네요. check style 돌려보셨나요?

thinker-94 replied:
네 Google Check style로 확인했습니다. 그런데 개행은 document에 있지만 ide 상으로는 체크를 해주지 않습니다. 그래서 document에서는 개행하는 방법이 여러가지 있지만 제가 적용하면 좋겠다고 생각한건 컴마를 기준으로 개행하는것입니다. 그리고 가독성을 위해 dto,enum,string 따로 나눠서 개행했습니다.

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -25,13 +26,16 @@
-    bookDTO.setSortTypeEnum(sortTypeEnum);
-
-    return bookRepository.selectBookByCategory(bookDTO);
+//  @Cacheable(cacheNames = "BookByCategoryCache", key = "{#categoryName}")
```
주석을 커밋하는것보단 그냥 해당 라인을 지우고 커밋해주세요

f-lab-dev reviewed (commented):
커밋메세지는 지금 사용하신 스타일 말고도 더 있을 겁니다. 여러가지로 찾아봐주세요.
그리고 커밋 제목에는 이슈번호를 달아주시면 좋겠습니다~

thinker-94 committed:
spring 의존하는 class를 제외 하고 모든 class를 final로 변경 [Issue #10]

스프링에서 호출되고 있는 클래스들을 제외하고 모든 클래스들을 final을 달아줌. 객체의 데이터가최대한 바뀌지 않게 하려면 클래스가 상속되지 않도록 해야되기 때문이다. 그리고 spring은 cglib를 통해서 클래스를 상속한뒤 프록시를 생성하기 때문에 스프링에 의존하는 클래스들은 final을 붙일 수 없다. 만약 클래스에 상속이 필요하다면 그때 final을 제거한다. 클래스를 상속으로는 일반적으로 만들지 않다고 생각하기 때문이다

thinker-94 committed:
BookService.getBookByCategory 줄바꿈 변경, @Cacheable 주석 제거 [Issue #10]

개행은 document에 있지만 ide 상으로는 체크를 해주지 않는다. 그래서 document에서는 개행하는 방법이 여러가지 있는걸 확인했다. 적용한건  컴마를 기준으로 개행하는 것이다. 그리고 가독성을 위해 dto,enum,string 따로 나눠서 개행했다

f-lab-dev commented:
커밋 메세지를 공부노트로 쓰시지말고 작업내용을 나타내는데 써주세요. 그리고 #10 에서 이슈라는게 유추가 됩니다. issue라는걸 굳이 써줄 필요가 있을까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/dto/PagingDto.java:
```diff
@@ -5,7 +5,10 @@
 @Setter
-public final class PagingDto {
+/*
+ * 해당 class는 @setter 때문에 불변 객체가 아니다.
```
어떤것때문에 불변이 아니게 만든걸까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/service/BookService.java:
```diff
@@ -25,17 +26,15 @@
   */
 
-  //TODO. key 추가로 만들것
-//  @Cacheable(cacheNames = "BookByCategoryCache", key = "{#categoryName}")
```
이 캐시는 왜 제거하셨을까요?

thinker-94 replied:
제거하지 않았습니다. 주석 했다가 주석 없앴습니다. ㅎ

f-lab-dev replied:
코드상으로 봤을때는 캐시 어노테이션이 제거되었으니 캐시기능도 적용이 안될 것 같은데요~

thinker-94 replied:
제가 커밋을 잘못했네요 캐시 어노테이션 지우고 다시 추가 하였습니다.

![image](https://user-images.githubusercontent.com/63249033/90447391-2852b980-e11e-11ea-967b-5e26e51b71de.png)

thinker-94 reviewed (commented):
> 커밋 메세지를 공부노트로 쓰시지말고 작업내용을 나타내는데 써주세요. 그리고 #10 에서 이슈라는게 유추가 됩니다. issue라는걸 굳이 써줄 필요가 있을까요?

넵 작업내용을 적겠습니다

thinker-94 committed:
[#10] PagingDto 클래스가 불변이 아닌 이유 주석으로 추가

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/dto/PagingDto.java:
```diff
@@ -6,7 +6,8 @@
 /*
- * 해당 class는 @setter 때문에 불변 객체가 아니다.
+ * 해당 class는 @Setter 때문에 불변 객체가 아니다.
+ * @Setter는 setter함수를 public으로 만들기 때문에 setter가 적용된 인스턴스 변수를 다른 사용자가 변경할 수 있게 한다.
```
이거는 이 클래스가 immutable이 아닌 이유가 아니라 단순히 immutable이 아닐때 어떻게 동작하는지를 설명한 것으로 보입니다.
이 클래스의 역할에 맞는 이유를 써주세요~

thinker-94 committed:
[#10] PagingDto class 역할 주석 작성
